internal/mesh/internal/controllers/routes/controller_test.go

```diff
@@ -0,0 +1,1033 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package routes
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	svctest "github.com/hashicorp/consul/agent/grpc-external/services/resource/testing"
+	"github.com/hashicorp/consul/internal/catalog"
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/mesh/internal/types"
+	rtest "github.com/hashicorp/consul/internal/resource/resourcetest"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"github.com/hashicorp/consul/proto/private/prototest"
+	"github.com/hashicorp/consul/sdk/testutil"
+	"github.com/hashicorp/consul/sdk/testutil/retry"
+)
+
+type controllerSuite struct {
+	suite.Suite
+
+	ctx    context.Context
+	client *rtest.Client
+	rt     controller.Runtime
+}
+
+func (suite *controllerSuite) SetupTest() {
+	suite.ctx = testutil.TestContext(suite.T())
+	client := svctest.RunResourceService(suite.T(), types.Register, catalog.RegisterTypes)
+	suite.rt = controller.Runtime{
+		Client: client,
+		Logger: testutil.Logger(suite.T()),
+	}
+	suite.client = rtest.NewClient(client)
+}
+
+func (suite *controllerSuite) TestController() {
+	// TODO: tidy comment
+	//
+	// This test's purpose is to exercise the controller in a halfway realistic
+	// way.
+
+	// Run the controller manager
+	mgr := controller.NewManager(suite.client, suite.rt.Logger)
+	mgr.Register(Controller())
+	mgr.SetRaftLeader(true)
+	go mgr.Run(suite.ctx)
+
+	backendName := func(name, port string) string {
+		return fmt.Sprintf("catalog.v1alpha1.Service/default.local.default/%s?port=%s", name, port)
+	}
+
+	var (
+		apiServiceRef = rtest.Resource(catalog.ServiceType, "api").Reference("")
+		fooServiceRef = rtest.Resource(catalog.ServiceType, "foo").Reference("")
+		barServiceRef = rtest.Resource(catalog.ServiceType, "bar").Reference("")
+
+		computedRoutesID = rtest.Resource(types.ComputedRoutesType, "api").ID()
+	)
+
+	// Start out by creating a single port service and let it create the
+	// default computed routes for tcp.
+
+	apiServiceData := &pbcatalog.Service{
+		Workloads: &pbcatalog.WorkloadSelector{
+			Prefixes: []string{"api-"},
+		},
+		Ports: []*pbcatalog.ServicePort{
+			{TargetPort: "tcp", Protocol: pbcatalog.Protocol_PROTOCOL_TCP},
+			{TargetPort: "mesh", Protocol: pbcatalog.Protocol_PROTOCOL_MESH},
+			// {TargetPort: "http", Protocol: pbcatalog.Protocol_PROTOCOL_HTTP},
+		},
+	}
+
+	_ = rtest.Resource(catalog.ServiceType, "api").
+		WithData(suite.T(), apiServiceData).
+		Write(suite.T(), suite.client)
+
+	var lastVersion string
+	testutil.RunStep(suite.T(), "default tcp route", func(t *testing.T) {
+		// Check that the computed routes resource exists and it has one port that is the default.
```

Comment:
Yes. You get a synthetic default route, which means that downstream consumers can always expect to get a non-nil CR, and can wait until it exists. That way the consumer doesn't have to figure out what it means to route traffic in the "default" way.

Reply:
> Related, when will the data in computed routes be nil? 
It will not be nil. If you switch from mesh to non-mesh the controller deletes the resource.